### PR TITLE
Protected containers

### PR DESCRIPTION
**Pull request by quvide**
# Protected containers
Add support for password protected containers. When creating a new instance, especially a Docker instance, the developer can optionally set a password in the creation dialog. If the password is non-empty, the user will be requested for the password before being granted access to the container.

After an instance has been created, its privacy settings cannot be changed. That is, the password protection cannot be removed, the password can't be removed, and if the instance is unprotected, a password can't be added. 

### Technical implementation
When creating the instance, the password is salted and hashed using bcrypt. The hash is then stored in the database in the `instances` table in the `blob` field.

Every time the instance is accessed, the proxy checks if the user is present in the `instance_auth` table along with the instance name. If the user is authorized, the proxy forwards the request normally. If it isn't present, the proxy instead presents a separate page to request the password from the user. If the user inputs the correct password, the page is reloaded and the user gets **permanent** access to the container.

### Api changes
* `POST /api/instance/new`: Added new parameter, `password` to the endpoint.
* `POST /api/authorization`: New endpoint which authorizes a user to access a container. See details in the API documentation.

### Database changes
New table called `container_auth`. The table contains `(user_id, instance_name)` pairs, which indicate that the user is authorized to access the instance.

---

## Review Checklist

- [x] All changes are described in the PR description
- [x] New feature and breaking changes have documentation
- [x] New endpoints have API tests
- [x] New frontend features have Enzyme tests
- [x] No unnecessary or debugging code
- [x] Questionable code is clearly marked with comments

### CODE REVIEW

**Comment by quvide** [comment]
@HenrikHelenius make `client/src/authorization-view.js` beautiful please

File: client/src/authorization-view.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  return (
+    <div className={css('auth-container')}>
+      <div className={css('dialog')}>
+        <h1>Authorization</h1>
+        <p>You are not authorized to access the container <code>{subdomain}</code></p>
+        <div className={css('form')}>
+          <input type="password" placeholder="Password" id="password" />
+          <button type="button" onClick={submitAuthTry}>Login</button>
```
Comment (kalomak) [inline client/src/authorization-view.js]: input type=submit migth be better

**Review by kalomak — DISMISSED** [review]
Looks great haha

File: client/src/authorization-view.js (hunk 1 above)
Comment (HenrikHelenius) [reply client/src/authorization-view.js]: yes change to that

Comment (quvide) [reply client/src/authorization-view.js]: why? preventDefault is not nice :(

**Review by kalomak — DISMISSED** [review]

File: client/src/authorization-view.js (hunk 1 above)
Comment (quvide) [reply client/src/authorization-view.js]: ok changed

**Review by kalomak — DISMISSED** [review]
👌

**Review by bqqbarbhg — APPROVED** [review]
⛵️⛵️⛵️⛵️

**Comment by quvide** [comment]
SHIPIT :shipit: :boat: :ship: :rocket:

**Pull request merged**